apps/sms/views/shared/js/navigation.js

```diff
@@ -1,361 +1,921 @@
-      var promise = Promise.resolve();
+  newPanelElement.classList.remove('panel-prepare-activation');
 
-      if (currentPanelObject && currentPanelObject.afterLeave) {
-        promise = promise.then(
-          currentPanelObject.afterLeave.bind(currentPanelObject, transitionArgs)
-        ).catch(catchError);
-      }
+  var doSlideAnimation =
+    from && from.panel && newView &&
+    from.panel.behavior !== newView.behavior;
 
-      if (nextPanelObject.afterEnter) {
-        promise = promise.then(
-          nextPanelObject.afterEnter.bind(nextPanelObject, transitionArgs)
-        ).catch(catchError);
-      }
+  console.log('doSlideAnimation', doSlideAnimation);
+  if (doSlideAnimation) {
+    newPanelElement.classList.add('panel-will-activate', 'panel-active');
+    oldPanelElement.classList.add('panel-will-deactivate');
+
+    newPanelElement.classList.toggle('panel-animation-reverse', isGoingBack);
+    oldPanelElement.classList.toggle('panel-animation-reverse', isGoingBack);
 
-      return promise;
+    return waitForSlideAnimation(newPanelElement).then(() => {
+      oldPanelElement.classList.remove(
+        'panel-will-activate', 'panel-will-deactivate',
+        'panel-active', 'panel-animation-reverse'
+      );
+      newPanelElement.classList.remove(
+        'panel-will-deactivate', 'panel-will-activate',
+        'panel-animation-reverse'
+      );
+    }, () => {
+      oldPanelElement.classList.remove(
+        'panel-will-deactivate', 'panel-animation-reverse'
+      );
+      newPanelElement.classList.remove(
+        'panel-will-activate', 'panel-animation-reverse'
+      );
     });
+  }
 
-    promise = promise.then(
-      function resolved() {
-        this.transitionPromise = null;
-        if (nextPanelContainer === currentPanelContainer) {
-          return;
-        }
-        nextPanelContainer.setAttribute('aria-hidden', false);
-        if (currentPanelContainer) {
-          currentPanelContainer.setAttribute('aria-hidden', true);
-        }
-      }.bind(this),
-      function rejected(e) {
-        catchError(e);
-        this.transitionPromise = null;
-        return Promise.reject(new Error('Error while transitioning'));
-      }.bind(this)
-    );
+  if (oldView) {
+    oldPanelElement.classList.remove('panel-active');
+  }
 
-    promise.
-      then(() => this.emit('navigated', { panel: panel, args: args })).
-      then(nextQueuedPanel, nextQueuedPanel);
+  if (newView) {
+    newPanelElement.classList.add('panel-active');
+  }
 
-    return (this.transitionPromise = promise);
-  },
+  return Promise.resolve();
+}
+
+/**
+ * Reacts to hash changes.
+ *
+ * This is the heart of the navigation logic and handles running the various
+ * navigation steps.
+ * It also handles any unwanted errors. (TODO: needs to tune this a little
+ * more).
+ */
+function onHashChange() {
+  debug('onHashChange, location=', window.location);
+  var newPanel = findViewFromLocation(window.location);
```

Comment:
__question:__ wondering if it's beneficial to move this piece of code to a separate function and reuse in ```init`` methods, something like:
```
function startNavigationFromLocation(location) {
	var newPanel = findViewFromLocation(location);
	if (!newPanel) {
		console.error('Couldn\'t find a view from location', location);
		return Promise.resolve();
	}

	var args = Utils.params(decodeURIComponent(location.hash));
	return startNavigation({ newPanel, args })
}
```

What do you think?